grin/src/Pipeline/Definitions.hs

```diff
@@ -0,0 +1,180 @@
+  | SparseCaseOptimisation
+  | UpdateElimination
+  | NonSharedElimination
+  | CopyPropagation
+  | ConstantPropagation
+  | DeadCodeElimination
+  | DeadDataElimination
+  | DeadFunctionElimination
+  | DeadParameterElimination
+  | DeadVariableElimination
+  | SimpleDeadFunctionElimination
+  | SimpleDeadVariableElimination
+  | SimpleDeadParameterElimination
+  | CommonSubExpressionElimination
+  | CaseCopyPropagation
+  | CaseHoisting
+  | GeneralizedUnboxing
+  | ArityRaising
+  | LateInlining
+  deriving (Enum, Eq, Ord, Show)
+
+newtype Hidden a = H a
+
+instance Show (Hidden a) where
+  show _ = "(hidden)"
+
+instance Eq (Hidden a) where
+  _ == _ = True
+
+data AbstractComputationStep
+  = CompileToAbstractProgram
+  | OptimiseAbstractProgram
+  | PrintAbstractProgram
+  | RunAbstractProgramPure
+  | PrintAbstractResult
+  deriving (Eq, Show)
+
+data SharingAnalysisStep
+  = ComputeSharing 
+  | PrintSharingResult
+  deriving (Eq, Show)
+
+data EffectStep
+  = CalcEffectMap
+  | PrintEffectMap
+  deriving (Eq, Show)
+
+data PipelineStep
+  = Optimize
+  | HPT AbstractComputationStep
+  | CBy AbstractComputationStep
+  | LVA AbstractComputationStep
+  | Sharing SharingAnalysisStep
+  | RunCByWithLVA
+  | Eff EffectStep
+  | T Transformation
+  | Pass [PipelineStep]
+  | PrintGrinH (Hidden (Doc -> Doc))
+  | PureEval
+  | JITLLVM
+  | PrintAST
+  | SaveLLVM Bool FilePath
+  | SaveGrin FilePath
+  | DebugTransformationH (Hidden (Exp -> Exp))
+  | Statistics
+  | ParseTypeAnnots
+  | PrintTypeAnnots
+  | PrintTypeEnv
+  | SaveTypeEnv
+  | Lint
+  | ConfluenceTest
+  | PrintErrors
+  | DebugPipelineState
+  deriving (Eq, Show)
+
+pattern PrintGrin :: (Doc -> Doc) -> PipelineStep
+pattern PrintGrin c <- PrintGrinH (H c)
+  where PrintGrin c =  PrintGrinH (H c)
+
+pattern DebugTransformation :: (Exp -> Exp) -> PipelineStep
+pattern DebugTransformation t <- DebugTransformationH (H t)
+  where DebugTransformation t =  DebugTransformationH (H t)
+
+data PipelineOpts = PipelineOpts
+  { _poOutputDir   :: FilePath
+  , _poFailOnLint  :: Bool
+  , _poLogging     :: Bool
+  , _poSaveTypeEnv :: Bool
+  , _poStatistics  :: Bool
+  }
+
+defaultOpts :: PipelineOpts
+defaultOpts = PipelineOpts
+  { _poOutputDir   = ".grin-output"
+  , _poFailOnLint  = True
+  , _poLogging     = True
+  , _poSaveTypeEnv = False
+  , _poStatistics  = False
+  }
+
+type PipelineM a = ReaderT PipelineOpts (StateT PState IO) a
+data PState = PState
+    { _psSrc           :: Maybe Text
+    , _psExp           :: Exp
+    , _psTransStep     :: Int
+    , _psSaveIdx       :: Int
+    , _psHPTProgram    :: Maybe HPTProgram
+    , _psHPTResult     :: Maybe HPTResult
+    , _psCByProgram    :: Maybe CByProgram
+    , _psCByResult     :: Maybe CByResult
+    , _psLVAProgram    :: Maybe LVAProgram
+    , _psLVAResult     :: Maybe LVAResult
+    , _psSharingResult :: Maybe SharingResult
+    , _psTypeEnv       :: Maybe TypeEnv
+    , _psTypeAnnots    :: Maybe TypeEnv
```

Comment:
Why do we need to maintain two type envs?

Reply:
Please write a comment about its meaning

Reply:
Done